docs/design_proposals/sync-improvements.md

```diff
@@ -0,0 +1,265 @@
+- `index.html` -> `/var/www/index.html`
+
+A fundamental change compared to the current sync mechanism is that one source path may by sync'ed to multiple destinations in the future.
+
+### Assumptions
+
+Builders know where to put static files in the container.
+This information needs to be pulled from the builders and made available in Skaffold.
+Currently, we know how to do this for Docker artifacts (@r2d4 suggested this in #1166).
+Whether jib and bazel builders can provide that information too, is unclear at the moment.
+
+### Interface changes
+
+There are at least two relevant changes
+
+1. When builders are queried for their dependencies, they must provide a list of destinations together with local source paths. The signature of `Builder.GetDependencies` (pkg/skaffold/build/build.go) therefore needs to change to:
+   ```go
+   type Builder interface {
+       Labels() map[string]string
+
+       Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact) ([]Artifact, error)
+
+       // OLD
+       // DependenciesForArtifact(ctx context.Context, artifact *latest.Artifact) ([]string, error)
+       // NEW
+       DependenciesForArtifact(ctx context.Context, artifact *latest.Artifact) (map[string][]string, error)
+   }
+   ```
+   
+2. To support additional sync maps from builders, the builder interface will need an additional method `SyncMap`:
+   ```go
+   type Builder interface {
+       Labels() map[string]string
+
+       Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact) ([]Artifact, error)
+
+       DependenciesForArtifact(ctx context.Context, artifact *latest.Artifact) (map[string][]string, error)
+    
+       // ADDED
+       SyncMap() map[string]string
+   }
+   ```
+
+### Config changes
+
+The `artifact.sync` config needs to support additional requirements:
+
+- Support existing use-cases for backwards compatibility, i.e.
+  - Glob patterns on the host path
+  - Transplant a folder structure, or flatten the folder structure
+- Offer a way to use destination inference.
+- Avoid ambiguity
+
+#### Problems with existing syntax
+- The flattening of the file structure at the destination is surprising. For example:
+  ```yaml
+  sync:
+    'src/**/*.js': dest/
+  ```
+  will put all `.js` files in the `dest` folder.
+- The triple-star syntax is is quite implicit and a surprising extension of the standard globbing syntax. For example:
+  ```yaml
+  sync:
+    'src/b/c/***/*.py': dest/
+  ```
+  will recreate subfolders _below_ `src/b/c` at `dest/` and sync `.py` files.
+  The triple-star therefore triggers two actions which should not depend on each other:
+  - do not flatten the directory structure
+  - strip everything on the source path up to `***`
+
+#### Suggestion
+These above problems can be made less surprising by a more explicit configuration syntax.
+The `artifact.sync` will be upgraded to a list of *sync rules*.
+
+```yaml
```

Comment:
This is amazing! I think we should definitely get other's input in this and discuss in the biweekly.